Update readme to include use cases and alternatives (#39)
XenitAB/spegel: README.md

Spegel, mirror in Swedish, is a stateless cluster local OCI registry mirror.

<p align="center">
  <img src="./assets/overview.gif">
</p>

## Use Cases

> Spegel is for you if you are looking to do any of the following.

Reviewer: ```suggestion
Spegel is for you if you are looking to do any of the following:
```

* Locally cache images from external registries with no explicit configuration.
* Avoid cluster failure during external registry downtime.
* Improve image pull speed and pod startup time by pulling images from the local cache first.

> * Avoid rate-limiting when pulling images from Docker Hub.

Reviewer: ```suggestion
* Avoid rate-limiting when pulling images from external registries (e.g. Docker Hub).
```

* Decrease network bandwidth usage outside of the private network.
* Increase image pull efficiency in edge node deployments.

## Background

Kubernetes does a great job at distributing workloads on multiple nodes. Allowing node failures to occur without affecting uptime. A critical component for this to work is that each node has to be able to pull the workload images before they can start. Each replica running on a node will incur a pull operation. The images may be pulled from geographically close registries within the cloud provider, public registries, or self-hosted registries. This process has a flaw in that each node has to make this round trip separately. Why can't the nodes share the image among themselves?

> Spegel enables all nodes in a Kubernetes cluster to act as a local registry mirror, allowing nodes to share images between themselves. Any image already pulled by a node will be available for any other node in the cluster to pull.

Reviewer: ```suggestion
Spegel enables each node in a Kubernetes cluster to act as a local registry mirror, allowing nodes to share images between themselves. Any image already pulled by a node will be available for any other node in the cluster to pull.
```

This has the benefit of reducing workload startup times and egress traffic as images will be stored locally within the cluster. On top of that it allows the scheduling of new workloads even when external registries are down.

> ## Prerequisite

Reviewer: ```suggestion
## Prerequisites
```
maybe?

Spegel relies on [Containerd registry mirroring](https://github.com/containerd/containerd/blob/main/docs/hosts.md#cri) to route requests to the correct destination. It is capable of writing the mirror configuration on startup, as it is loaded on the fly by Containerd.
It is however not capable of setting the registry mirror `config_path` in the Containerd configuration as it requires a restart of Containerd for the change to be picked up.

> Spegel has been verified with AKS and EKS as they come with the expected configuration out of the box. Any other Kuberentes deployment should make the following configuration change to Containerd.

Reviewer: ```suggestion
Spegel has been verified with AKS and EKS as they come with the expected configuration out of the box. Any other Kubernetes deployment should make the following configuration change to Containerd:
```

In its core Spegel is a pull only OCI registry which runs locally on every Node in the Kubernetes cluster. Containerd is configured to use the local registry as a mirror, which would serve the image from within the cluster or from the source registry.

## Alternatives

### Private Registry 

> A common practice, especially for larger enterprises, is to run a private registry like Harbor to replica images from public registries. Storing them within the private network close to the cluster.

Reviewer: ```suggestion
A common practice, especially for larger enterprises, is to run a private registry like Harbor to replicate images from public registries, storing them within the private network close to the cluster.
```

This is a great option for those who have the time and budget to invest in running and managing the infrastructure. For others, it may be a good practice but unattainable in reality.
Spegel does not aim to replace projects like [Harbor](https://github.com/goharbor/harbor) or [Zot](https://github.com/project-zot/zot) but instead complements them. Having a persistent copy of public images stored geographically close to a cluster is great. Spegel will however enable

> nodes to pull from images closer as long as the images are somewhere within the cluster. Additionally, there is no guarantee that a self-managed private registry is never unavailable. In these scenarios

Reviewer: ```suggestion
nodes to pull from images closer as long as the images are somewhere within the cluster. Additionally, there is no guarantee that a self-managed private registry is always available. In these scenarios
```
remove double negation

> running Spegel is like having both belt and braces.

Reviewer: ```suggestion
running Spegel is like wearing both belt and suspenders.
```

### Dragonfly

[Dragonfly](https://github.com/dragonflyoss/Dragonfly2) is a great project that has been around for a while. In some aspects, Spegel takes inspiration from the work done by Dragonfly. 

> The difference is that Spegel aims to solve a smaller problem set. While it may mean fewer features it also means fewer moving components. Dragonfly requires both Redis and Mysql which 

Reviewer: ```suggestion
The difference is that Spegel aims to solve a smaller problem set. While it may mean fewer features it also means fewer moving components. Dragonfly requires both Redis and MySQL which 
```

increases the resource consumption and burden on end users to manage additional resources. It also increases the risk of errors occurring during critical moments. The benefit of Spegel
is that it is stateless meaning that any temporary failure of nodes and communication should be easily resolved automatically.

### Kraken

> [Kraken](https://github.com/uber/kraken) implements a similar solution to Spegel with its P2P agent component. It is however not heavily maintained, meaning that new features and security updates will not be added.

Reviewer: Not sure how "heavily" should be interpreted? It appears to be maintained, in that it had a commit in January, but no release since 2020. It certainly appears to be "maintained", but maybe not "actively" maintained. Anyway, I don't have a better suggestion than "heavily", I'm just not sure how a reader should understand that choice of word.

Author: I don't know myself, as there seems to be some activity but no real community engagement looking at the issues. I will leave this for now until we come up with something better.
